packages/plugin-core/src/commands/MoveHeader.ts

```diff
@@ -0,0 +1,311 @@
+      HistoryService.instance().subscribev2("lookupProvider", {
+        id: this.key,
+        listener: async (event) => {
+          if (event.action === "done") {
+            HistoryService.instance().remove(this.key, "lookupProvider");
+            const cdata = event.data as NoteLookupProviderSuccessResp;
+            resolve({
+              dest: cdata.selectedItems[0],
+              origin: maybeNote,
+              nodesToMove,
+              modifiedOriginTree: tree,
+              originProc: proc,
+              engine,
+            });
+            lookupController.onHide();
+          } else if (event.action === "error") {
+            this.L.error({ msg: `error: ${event.data}` });
+            resolve(undefined);
+          } else if (event.action === "changeState") {
+            this.L.info({ msg: "cancelled" });
+            resolve(undefined);
+          } else {
+            this.L.error({ msg: `unhandled error: ${event.data}` });
+            resolve(undefined);
+          }
+        },
+      });
+    });
+  }
+
+  removePosition = (nodes: Node[]) => {
+    return _.map(nodes, (node) => {
+      visit(node, (n) => {
+        delete n["position"];
+      });
+      return node;
+    });
+  };
+
+  async execute(opts: CommandOpts): Promise<CommandOutput> {
+    const ctx = "MoveHeaderCommand";
+    this.L.info({ ctx, opts });
+    const {
+      dest,
+      origin,
+      nodesToMove,
+      modifiedOriginTree,
+      originProc,
+      engine,
+    } = opts;
+
+    // deep copy the origin before mutating it
+    const originDeepCopy = _.cloneDeep(origin);
+    // remove header from origin
+    const modifiedOriginContent = originProc.stringify(modifiedOriginTree);
+
+    origin.body = modifiedOriginContent;
+    await engine.writeNote(origin, {
+      updateExisting: true,
+    });
+
+    // append header to destination
+    const destProc = this.getProc(engine, dest!);
+    const destTree = destProc.parse(dest!.body);
+    destTree.children = (destTree.children as Node[]).concat(nodesToMove);
+    const modifiedDestContent = destProc.stringify(destTree);
+    dest!.body = modifiedDestContent;
+    await engine.writeNote(dest!, {
+      updateExisting: true,
+    });
+
+    delayedUpdateDecorations();
+
+    // update reference
+    const anchorsBefore = RemarkUtils.findAnchors(originDeepCopy.body);
+    const anchorsAfter = RemarkUtils.findAnchors(modifiedOriginContent);
+    const anchorsToUpdate = _.differenceWith(
+      anchorsBefore,
+      anchorsAfter,
+      this.hasIdenticalChildren
+    );
+    const anchorsToUpdateNames = _.map(anchorsToUpdate, (anchor) => {
+      const slugger = getSlugger();
+      const payload = AnchorUtils.anchorNode2anchor(anchor, slugger);
+      return payload![0];
+    });
+
+    const foundReferences = await findReferences(origin.fname);
+    const { vaults, wsRoot, notes } = engine;
+    const updated: NoteProps[] = [];
+    _.forEach(foundReferences, async (reference) => {
```

Comment:
Also discussion point: we often use lodash methods when streaming API has an alternatives. 

As example this code could be written with streaming api such that there is less state that we are dealing with while streaming through the data, with less state at each step in theory should be much easier to grasp the logic. 

```
    foundReferences
      .filter(r => !r.isCandidate)
      .map(r => this.getNoteByLocation(r.location.uri.fsPath, wsRoot, vaults, notes))
      .filter(note => note !== undefined)
      .forEach(note => {
        // find match text in note and modify the node
        const proc = this.getProc(engine, note!);
        const tree = proc.parse(note!.body);
        visit(tree, (node: Node) => {
          if (node.type === DendronASTTypes.WIKI_LINK) {
            if (
              node.value === origin.fname &&
              _.includes(anchorsToUpdateNames, node.data!.anchorHeader)
            ) {
              node.value = dest!.fname;
            }
          }
        });
        note!.body = proc.stringify(tree);
        const writeResp = await engine.writeNote(note!, { updateExisting: true });
        updated.push(writeResp.data[0].note);
      });

```

Reply:
I think this is another coding style (or the lack thereof) issue. We try to use lodash where using it makes things easier, but we don't really have a concrete guiding principle for what you describe here. I myself go back and forth using lodash native js methods with no solid reasoning as to why I'm doing that. It's definitely a good topic to discuss. (probably not here)

Reply:
Went with the suggestion here as well.